blog/2023-02-16-do-you-need-to-build-auth/index.md

```diff
@@ -0,0 +1,124 @@
+
+> **TL;DR** If you're still learning, auth will be a good challenge. If you want full control, it's worth building your own auth system. Otherwise, it depends on several conditions.
+
+## Intro
+
+As a developer, I have built many applications during my career. No matter the programming language, there is a common foundation I always need to construct: user auth.
+
+It was a negligible part since everything was straightforward going back to 20 years ago:
+
+- Implement a registration and sign-in system with username and password.
+- Create a mechanism to maintain a user's session.
+- About security? MD5 is the answer.
+
+That’s it. Then I could focus on “the real business”. Didn't hear about MD5? You missed out on the "good times" of software development. Nowadays, developers face more challenges when building sign-in/up.
+
+It sounds alarmist, but let me go through with an example.
+
+## Example: An online bookstore
+
+Let’s say you are trying to build an online bookstore with an API service and a web frontend app.
+
+First, the scope of “auth” should be defined. Auth can be explained as authentication and authorization, and they have totally different definitions and use cases:
+
+:::note 🤔 About Auth
+
+- Authentication (AuthN) answers the question “Which identity do you own?”
+- Authorization (AuthZ) answers the question “What can you do?”
+  :::
+
+I wrote an article [CIAM 101: Authentication, Identity, SSO](../2022-11-28-ciam-101-authn-identity-sso/index.mdx) to discuss these concepts in detail.
+
+### Choose authentication methods
+
+Let’s start with “authentication”, which is user sign-in in our example. Besides the username-and-password method, here are some trending methods people are adopting for a better user conversion and security:
+
+- Passwordless, i.e. dynamic verification code (via email or sms)
+- Social sign-in (Google, GitHub, etc.)
+
+The choice of method depends on the business decision, while we can take a look on the tech effort:
+
+- For passwordless, you need to find a vendor to send email or sms; then integrate with your API service (new APIs may needed).
+- For social sign-in, you must adhere to the integration guidelines of the social identity provider(s) you wish to use. Additionally, you must create a mapping between your bookstore's user IDs and the identity provider's.
+  - For example, when a user signs in with a Gmail account `foo@gmail.com`, you need to link this address to the user `foo` in the bookstore. This is to implement a unified identity system and give the user the ability to change or unlink their social identity in the future.
+
+### Design and implement sign-in experience
+
+After you decide authentication methods, a graceful and smooth sign-in experience for your end-users is the next target. The conversion here is fundamental but crucial to the business since it's a nature way to leave the persist customer data.
```

Comment:
```suggestion
After you decide authentication methods, a graceful and smooth sign-in experience for your end-users is the next target. The conversion here is fundamental but crucial to the business since it's a natural way to leave the persisted customer data.
```